### DISCUSSION TITLE
Plug In Not Showing in OBS 28.0.3 [MacOS x86]

### DISCUSSION THREAD

**Discussion by djSmileyMike** [WarmUpTill/SceneSwitcher#567] [Q&A]
Hi, 

I recently upgraded to OBS 28.0.3 and tried reinstalling Advanced Scene Switcher, but it's not showing up in the Tools menu (or anywhere else that I can see). I've tried running the universal and the x86 installer. And I tried granting full disk privileges to the installer to see if that made a difference. 

It looks like ASS is not in the OBS package plugins folder when I open that. 

When I look in Library/Application Support/obs-studio/plugins/ I see advanced-scene-switcher.plugin but no /advanced-scene-switcher subfolder and no advanced-scene-switcher.so

I've tried expanding the installer package with kaka, but that gives me another package called: advanced-scene-switcher.pkg and when I show contents of that I don't get the advanced-scene-switcher.so and data folder that I need to try and do a manual installation. 

Please help!
Thanks!

**Comment by WarmUpTill** [comment]
Can you share an OBS log?

>It looks like ASS is not in the OBS package plugins folder when I open that.

It should not be added as part of the OBS.app itself.

>When I look in Library/Application Support/obs-studio/plugins/ I see advanced-scene-switcher.plugin but no /advanced-scene-switcher subfolder and no advanced-scene-switcher.so

That seems really strange.
You are referring to the ~/Library/... path here, correct?

**Reply by djSmileyMike** [reply]
I was referring to absolute /Library/Application Support/obs-studio/plugins/

~/Library/Application Support/obs-studio/plugins/ has nothing related to ASS in it. 

Note that screenshot: Screen Shot 2022-10-14 at 1.06.11 PM.png shows x86 installer struggling with permissions. (Mac's have gotten pretty tight on restricting folder access to applications. Sometimes it has to be specifically granted). So I suspect this might be relating to the installer not having permission to copy files into the right location. However, if I grant the installer full disk access then I can get past that error, but the files still don't seem to be in the right place. 

If I can get at the files I could try a manual install. Not sure if I'm uncompressing the package wrong, but I don't seem to be able to get to the data folder you mention in the manual install instructions. 

<img width="622" alt="Screen Shot 2022-10-14 at 1 06 11 PM" src="https://user-images.githubusercontent.com/115808059/195932981-7c9dfb20-0518-4e86-8f3f-a9288927e758.png">

[2022-10-14 13-00-23.txt](https://github.com/WarmUpTill/SceneSwitcher/files/9789717/2022-10-14.13-00-23.txt)

**Reply by WarmUpTill** [reply]
>Note that screenshot: Screen Shot 2022-10-14 at 1.06.11 PM.png shows x86 installer struggling with permissions. (Mac's have gotten pretty tight on restricting folder access to applications. Sometimes it has to be specifically granted). So I suspect this might be relating to the installer not having permission to copy files into the right location. However, if I grant the installer full disk access then I can get past that error, but the files still don't seem to be in the right place.

I see - which version of MacOS are you using?
I only have a VM on which I can test the MacOS version the plugin and that is still stuck on Big Sur.

>If I can get at the files I could try a manual install. Not sure if I'm uncompressing the package wrong, but I don't seem to be able to get to the data folder you mention in the manual install instructions.

I have attached the contents of the x86 pkg as a zip file.
[advanced-scene-switcher-macos-x86_64.zip](https://github.com/WarmUpTill/SceneSwitcher/files/9791699/advanced-scene-switcher-macos-x86_64.zip)

In my VM the files are installed at the following folder with "mac" being the username:

```
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/_CodeSignature
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/_CodeSignature/CodeResources
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/advanced-scene-switcher
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/adv-ss-plugins
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/adv-ss-plugins/libopencv_features2d.406.dylib
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/adv-ss-plugins/libopencv_objdetect.4.6.0.dylib
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/adv-ss-plugins/libopencv_calib3d.4.6.0.dylib
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/adv-ss-plugins/libopencv_calib3d.406.dylib
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/adv-ss-plugins/libopencv_core.406.dylib
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/adv-ss-plugins/libopencv_flann.406.dylib
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/adv-ss-plugins/libopencv_core.4.6.0.dylib
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/adv-ss-plugins/libopencv_imgproc.406.dylib
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/adv-ss-plugins/libopencv_features2d.4.6.0.dylib
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/adv-ss-plugins/libopencv_flann.4.6.0.dylib
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/adv-ss-plugins/libopencv_imgproc.4.6.0.dylib
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/adv-ss-plugins/advanced-scene-switcher-opencv.so
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/adv-ss-plugins/libopencv_objdetect.406.dylib
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/MacOS/advanced-scene-switcher-lib.so
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/res
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/res/time.svg
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/res/cascadeClassifiers
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/res/cascadeClassifiers/haarcascade_upperbody.xml
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/res/cascadeClassifiers/haarcascade_profileface.xml
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/res/cascadeClassifiers/haarcascade_frontalface_alt2.xml
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/res/cascadeClassifiers/haarcascade_eye.xml
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/res/cascadeClassifiers/haarcascade_lefteye_2splits.xml
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/res/cascadeClassifiers/haarcascade_frontalface_alt_tree.xml
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/res/cascadeClassifiers/haarcascade_righteye_2splits.xml
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/res/cascadeClassifiers/haarcascade_frontalface_alt.xml
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/res/cascadeClassifiers/haarcascade_lowerbody.xml
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/res/cascadeClassifiers/haarcascade_frontalface_default.xml
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/res/cascadeClassifiers/haarcascade_fullbody.xml
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/res/cascadeClassifiers/haarcascade_eye_tree_eyeglasses.xml
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/locale
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/locale/tr-TR.ini
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/locale/ru-RU.ini
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/locale/de-DE.ini
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/locale/es-ES.ini
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/locale/zh-CN.ini
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Resources/locale/en-US.ini
/Users/mac/Library/Application Support/obs-studio/plugins/advanced-scene-switcher.plugin/Contents/Info.plist
```

**Reply by djSmileyMike** [reply]
I'm on Catalina 10.15.7
Let me know if I can help you QA anything. 

I'll try a manual install and see if that works. 
I notice that your directory mapping above differs from the manual instal notes on this page: https://obsproject.com/forum/resources/advanced-scene-switcher.395/ 
Under Spoiler MacOS it mentions a data folder needing to be copied and that advanced-scene-switcher.so file should go in Library/Application Support/obs-studio/plugins/advanced-scene-switcher/bin/ and it doesn't mention anything about the advanced-scene-switcher.plugin file (which I think did install on my system, although it was in the root /Libaray/... path and not the user ~/Library path. 

I'll let you know what happens with my manual install attempt. 
Cheers!

**Reply by djSmileyMike** [reply]
Success! But with a little tweaking...

It does seem to be related to file permissions. 

Here's the steps that worked for me: 
1. Copy advanced-scene-switcher.plugin into ~/Library/Application Support/obs-studio/plugins/
2. Open ~/Library/Application Support/obs-studio/plugins/ in Terminal
3. Run command: xattr -r -d com.apple.quarantine advanced-scene-switcher.plugin

Here's what's happening:

Because you're not an identified developer by Apple, your plugin is getting quarantined. When an Application is from an unidentified developer and you try to run it, Apple warns you, and then you can go and allow an exception in the Security & Privacy preferences panel. However, because your software is a plugin and not an application, there's nowhere to override it, and the error messages it's throwing aren't a perfect match. (I'll attached some example screenshots). Running the command in #3 above manually removes the quarantine flag recursively on all files and folders in the plugin package. Now when you run OBS Advanced Scene Switcher loads in the Tools menu and the first time use pop-up appears. 

Screenshots: 
<img width="415" alt="quarantine-error-user-library" src="https://user-images.githubusercontent.com/115808059/195998002-61ef2709-0fe2-411d-a35e-850b4fb35861.png">
This is the error you get when the plugin is installed in ~/Library/Application Support/obs-studio/plugins/ and the quarantine flags haven't been cleared. Note that the plugin file is NOT damaged! This is just the closest error that OBS is able to throw. 

<img width="559" alt="extended-attribute-flags" src="https://user-images.githubusercontent.com/115808059/195998060-7efbd111-6be7-40f9-b73b-e9c8462b439b.png">
The @ symbol at the end of the file permissions indicates that advanced-scene-switcher.plugin has extended attributes. The "xattr advanced-scene-switcher.plugin" command lists the extended attributes and you can see "com.apple.quarantine" is listed. 

Guess on Installer Failing:
My guess is that your installer isn't recognized by Apple and so it's struggling with permissions (hence that initial error where it says it can't install to the current user and you can select anything else). When I grant the installer an exception, it goes ahead with the install, but puts the files in the root /Library/... path instead of the user ~/Library/... path. Even though the files get copied, they have the quarantine extended attribute set. I think OBS might be missing the plugin on load because it's in the root Library instead of the user Library (which is a silent fail) and even if I move it into user Library path it's still failing (less silently) because it's quarantined. 

Let me know if you want any assistance testing an update to the Installer. 
Great plug in by the way! It's absolutely essential to my stream! I was lost without it this week!

**Reply by WarmUpTill** [reply]
Thank you very much for tracking down the problem!
I will add notes about your findings to the installation section on the plugin overview page.

>I think OBS might be missing the plugin on load because it's in the root Library instead of the user Library (which is a silent fail) and even if I move it into user Library path it's still failing (less silently) because it's quarantined.

Both paths should be supported by OBS.
I moved the plugin to `~/` on request of other users rightfully complaining about the older installer requiring root privileges. (#373)
To be honest I am not sure what dictates whether `/` or `~/` is chosen

Unfortunately, being an "identified Apple developer" would cost me 99$ a year (At least that is what I heard), which seems a bit costly.
I am not sure how to avoid such issues.

**Reply by djSmileyMike** [reply]
The workaround manual install isn't very complicated. I didn't need any sudo privileges or anything fancy to do what I did. So if there's clear instructions for manual install and you provide that zip file like you did for me, I think most people should be able to figure out. (If they can figure out all the video/codec/bitrate configurations required for streaming, they should be able to handle these steps! 😂) 

I agree that using ~/ is likely better. The root /Library folder might require sudo to change permissions. However, note that when the plug in was in root /Library/... the OBS load failed silently. So if people are just getting a missing plugin with no error on load, they might have the plugin in the root /Library folder. OBS might scan both on load, but it doesn't seem to handle the error properly in either case.

**Comment by biggator** [comment]
I experienced the same issue and was lucky to find this page. I installed it manually into `~/Library/Application Support/obs-studio/plugins` with these instructions.

The .pkg installer won't run initially because it is not signed by a developer account, so I had to bypass the security by starting with `ctrl-click`. Then, when the installer completes, it announces success and yet does not actually install anything.

It would be nice if these instructions from @djSmileyMike were included in the README. Of course, we'd need access to the .zip files instead of the .pkg.

Let me know if I can help.

- macOS Ventura 13.1
- OBS Studio 28.1.2

**Reply by emmanueluk** [reply]
I'm struggling with this... I remember going through misery before.  Can you help me by writing down what I should do to get this plug in working.  OBS 29, Monterey  12.6.2.  I have tried for hours using terminal and copying the file  and folders but nothing.  I have downgraded to v27.  I've run out of options.  Help please :)

**Reply by jgwynngmail** [reply]
I'm having a similar issue to the above--plugin does not show after install.  I have the plugin in the directory(see screenshot), and used the command xattr -r -d com.apple.quarantine advanced-scene-switcher.plugin to remove the quarantine.  Running OBS 27.0.1 on Mac OS 10.14.6
<img width="702" alt="Screen Shot 2023-06-04 at 8 06 25 PM" src="https://github.com/WarmUpTill/SceneSwitcher/assets/135561011/26f57936-377a-4801-a05b-ad0a431921da">

**Reply by WarmUpTill** [reply]
Can you please share an OBS log?

**Reply by jgwynngmail** [reply]
Here is a recent log file.  Thanks for any advice/solution.
[2023-06-05 09-36-39.txt](https://github.com/WarmUpTill/SceneSwitcher/files/11664418/2023-06-05.09-36-39.txt)

**Reply by WarmUpTill** [reply]
You are using OBS 27.
Unfortunately this version is no longer supported, due to a dependency change about a year ago on OBS side.

If you can, please update to at least OBS 28.

If you have to stick with OBS 27, the last supported version of the plugin is this one:
https://github.com/WarmUpTill/SceneSwitcher/releases/tag/1.17.7
